File: Leetcode/Algorithm/python/0000/00222-Count Complete Tree Nodes.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def countNodes(self, root):
        if not root:
            return 0
        
        (ll, lr) = self.get_depth(root.left)
        (rl, rr) = self.get_depth(root.right)
        if ll == rr:
            return (2 ** (ll + 1)) - 1
        
        if ll == lr:
            a = (2 ** ll) - 1
            b = self.countNodes(root.right)
            return a + b + 1
        else:
            a = self.countNodes(root.left)
            b = (2 ** rr) - 1
            return a + b + 1
            
    def get_depth(self, root):
        l, r = 0, 0
        
        a = root
        while a:
            l += 1
            a = a.left
        
        b = root
        while b:
            r += 1
            b = b.right
        
        return (l, r)

File: Leetcode/Algorithm/cpp/0000/00222-Count Complete Tree Nodes.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int countNodes(TreeNode* root) {
        if (!root) {
            return 0;
        }
        int a = depthRight(root->left);
        int b = depthRight(root->right);
        
        if (a == b) {
            return (1 << b) + countNodes(root->left);
        } else if (a > b) {
            return (1 << a) + countNodes(root->right);
        }
    }
private:
    int depthRight(TreeNode* root) {
        int res = 0;
        while (root) {
            res++;
            root = root->right;
        }
        return res;
    }
};
